### ISSUE TITLE
support other protocols from matterbridge

### ISSUE THREAD

**Issue by Zauberfisch** [42wim/matterircd#266]
matterbridge seems to support a lot more protocols than matterircd. 
I'm sure there is a good reason for it, but much to my surprise I didn't see any mention about it in the readme or any other existing issues.

So please forgive my ignorance and let me ask, are there any plans to support more protocols?
And what would be needed to make that happen?

I realise that there are other existing solutions that attempt to do that, most notably, bitlbee. But for example the discord extension to bitlbee and libpurple both are terrible.

(keywords for other people searching: (Discord, Telegram, Rocket.chat, Matrix, Steam, Twitch, Ssh-chat, WhatsApp, Zulip, Keybase)

**Comment by paretje** [comment]
matterbridge recently gained support for Microsoft Teams. As there doesn't seem to be anything like matterircd supporting Microsoft Teams, or e.g. a CLI client, it would be really nice if this was added to matterircd. It also kind of fits well with the current support for Mattermost and Slack.

**Comment by Zauberfisch** [comment]
Hey @42wim, sorry to ping you directly, but I'd love to hear your input on this.

**Comment by 42wim** [comment]
@Zauberfisch  Well, for some protocols it would be possible, depending on how they support private messages.
But for now the main problem for me is time.
So in short term no other protocols will be added (unless other people make PR's)

**Comment by Zauberfisch** [comment]
@42wim fair enough. Time being limited is a problem I understand all to well. Thanks for all your great work.
But glad to know that you are open minded about it.

**Comment by Zauberfisch** [comment]
not directly related to matterircd, but a quick update for others looking for a discord<>irc adapter. I am now using [rdircd](https://github.com/mk-fg/reliable-discord-client-irc-daemon) which works well and has almost all features I need.

**Comment by codygman** [comment]
Can you add in the readme that matterircd doesn't support all of the services that matterbridge uses? I've spent quite a bit of time wondering why my discord<>irc bridge using matterbridge<>matterircd wasn't working and I'm betting others could fall into the same trap.

**Comment by Zauberfisch** [comment]
@codygman I'm not sure what you mean by that.

But I thought the docs are rather clear.
This project doesn't really have anything to do with matterbridge. And it's clearly stated that it only supports mattermost and slack.

discord < matterbridge > irc does works
discord < matterircd > irc does NOT work
discord < matterbridge > mattermost<matterircd>matterircd will also work

**Comment by 42wim** [comment]
Wrt to other protocols, I'm refactoring matterircd which allows much easier plugins of another protocol.
I'm also looking for testers for this refactoring, so those who want to help, please try out the refactor branch :-)

**Comment by Zauberfisch** [comment]
@42wim very nice. I'd be happy to test.

How "experimental" should I expect the refactor branch to be? 
Not sure if I should setup a 2nd instance of it, or just replace my current matterircd setup.

**Comment by 42wim** [comment]
@Zauberfisch planning on merging it into master shortly, (going to fix some open issues first) it's on the refactor2 branch now.
The basic code is almost the same (and also not), it should not be that experimental, but it's only tested by me for now.

**Comment by 42wim** [comment]
If anyone would want to take a crack at it, it should be possible now to plugin other bridges, by implementing the bridge interface
https://github.com/42wim/matterircd/blob/master/bridge/bridge.go for the protocol you want.

**Comment by Zauberfisch** [comment]
Would that be an api that is 1:1 compatible with matterbridge? Or will bridges to other protocols still need to be rewritten to fit into matterircd?

I haven't gotten around to testing the new branch yet, but I'm hoping I have some time this weekend

**Comment by 42wim** [comment]
need to be rewritten, matterbridge and matterircd are two quite different beasts

**Comment by Zauberfisch** [comment]
Figured, but worth asking :)

well, guess it's time I'll learn some go ...

**Comment by hloeung** [comment]
Wee, WIP Matrix support https://github.com/42wim/matterircd/pull/376
